### PR TITLE
Support Zcash v5 transaction format on TT

### LINKED ISSUE #2031: Support ZIP-244 TxIDs
The NU5 network upgrade in Zcash introduces a new TxID digest [1]. The legacy TxIDs will still be accepted by the network, but in order for Trezor to be able to spend UTXOs from transactions which use the new ZIP-244 digest, Trezor will apparently need to implement the new algorithm, so that it may verify previous transaction output amounts. The NU5 upgrade should activate on Zcash Mainnet on 18th April 2022 [2].

- In trezor-core the `Zcashlike` class will need to override `get_prevtx_output()` to compute the legacy TxID or ZIP-244 TxID depending on the transaction version. 
- The same will be needed in trezor-legacy.
- Updates may also be needed in trezor-suite and blockbook. Further investigation is needed.

[1] ZIP-244: https://zips.z.cash/zip-0244#txid-digest
[2] https://z.cash/support/schedule/

### PR DESCRIPTION
**Pull request by jarys**
Implements #2031 for Trezor T. (not for Trezor 1, so the this PR does not close the issue)

This PR implements

- new Zcash transaction format ([ZIP-225](https://zips.z.cash/zip-0225)) and
- new Zcash sighash algorithm ([ZIP-244](https://zips.z.cash/zip-0244))

to support Zcash Network Upgrade 5.

I moved Zcash code into the separated app to prepare it for future extensions (like unified addresses and shielded transactions).

I kept old Zcash code to preserve Zcash-like altcoins like Komodo and Koto.

### CODE REVIEW

**Comment by jarys** [comment]
blocked by missing unit tests

File: core/src/apps/zcash/sig_hasher.py
Comment (andrewkozlik) [inline core/src/apps/zcash/sig_hasher.py:150]: I think the comments should go on a separate line above the code, because the code formatter will break them (not only here). Run `make style` in the root directory to format the code.

File: core/src/apps/zcash/sig_hasher.py
Comment (andrewkozlik) [inline core/src/apps/zcash/sig_hasher.py:43]: I am not too fond of these `initialize` methods. I think the proper way is for this to go in `__init__()`. I assume the reason why it's not there is because `Signer.create_sig_hasher()` doesn't have the `tx` parameter, so one solution would be:
- Add the `tx` parameter to `create_sig_hasher()` and `__init__()` and to remove it from `hash341()` and `hash143()`.

However, the sighasher doesn't actually need to be initialized with the `tx` parameter. It is only used to precompute the header digest and some personalization string, which can both be computed in `signature_digest()` and `txid_digest()` and optionally cached to avoid recomputing. So another option is:
- Add the `tx` parameter to `signature_digest()` and `txid_digest()` and cache the header digest.

File: core/src/apps/zcash/sig_hasher.py
Comment (andrewkozlik) [inline core/src/apps/zcash/sig_hasher.py:170]: I think in the end the Zcash people decided on really using the scriptPubKey, not the scriptCode, didn't they?

**Review by andrewkozlik — COMMENTED** [review]
Some comments to start out the review.

File: core/src/apps/zcash/sig_hasher.py
Comment (andrewkozlik) [inline core/src/apps/zcash/sig_hasher.py:65]: Note that this is dead code. We actually don't use or need `txid_digest()` anywhere, so I am wondering whether to keep the code for it. The only place that I can think of where it would be used are replacement transactions, which we do not support for Zcash, because fee bumping is not an issue. Is there any other possible use?

Comment (andrewkozlik) [reply core/src/apps/zcash/sig_hasher.py:65]: I can imagine that at some point in the future we will want to display the TXID on the Trezor screen at the end of the signing process. So I don't mind keeping the code for future use, just wanted to open a discussion.

File: core/src/apps/zcash/sig_hasher.py
Comment (andrewkozlik) [inline core/src/apps/zcash/sig_hasher.py:33]: We have two classes named `ZcashSigHasher`, which is confusing, even though they are in different modules. The original one can be renamed to `Zip243SigHasher` or `ZcashLegacySigHasher`. The newer one can be renamed to `ZcashHasher` or `Zip244Hasher`, because it does more than just signature hashes (assuming we keep `txid_digest()`). There is also the question of whether we need it to be distinct from `ZcashExtendedSigHasher`.

In summary:
Change original `ZcashSigHasher` to `Zip243SigHasher` or `ZcashLegacySigHasher`.
Change new `ZcashSigHasher` to `Zip244Hasher` or `ZcashHasher` or merge with `ZcashExtendedSigHasher`.
Change `ZcashExtendedSigHasher` to `Zip244SigHasher` or `ZcashSigHasher`.

File: core/src/apps/bitcoin/sign_tx/__init__.py
Comment (andrewkozlik) [inline core/src/apps/bitcoin/sign_tx/__init__.py:80]: I take it that `zcashlike.py` is for the Komodo and Koto coins. Are these going to keep using the older transaction format? If so, then I think we should be calling this `zcash_legacy.py` or `zcash_v4.py`, because "zcashlike" would seem to be implementing a superset of Zcash functionality, which it will no longer be doing.

File: core/src/apps/zcash/signer.py
Comment (andrewkozlik) [inline core/src/apps/zcash/signer.py:79]: These do not need to be `async`.

File: core/src/apps/zcash/signer.py
Comment (andrewkozlik) [inline core/src/apps/zcash/signer.py:72]: Semantically I think this actually belongs somewhere else. It's much more than just the outputs, it's both spends and outputs or entire actions. I'd suggest putting it in `write_tx_footer()` or `step7_finish()` or maybe even `step6_sign_segwit_inputs()`?

File: core/src/apps/zcash/signer.py
Comment (andrewkozlik) [inline core/src/apps/zcash/signer.py:87]: `dump` -> `bump`

File: core/src/apps/zcash/signer.py
Comment (andrewkozlik) [inline core/src/apps/zcash/signer.py:63]: algortihm -> algorithm

File: tests/device_tests/zcash/test_sign_tx.py
Comment (andrewkozlik) [inline tests/device_tests/zcash/test_sign_tx.py:104]: Let's be sure that the transactions work on testnet, before merging the code.

File: tests/device_tests/zcash/test_sign_tx.py
Comment (andrewkozlik) [inline tests/device_tests/zcash/test_sign_tx.py:256]: It would be nice to have one multisig test.

**Review by andrewkozlik — CHANGES_REQUESTED** [review]
One last thing I am wondering about is the movement of the Zcash code up by one level. I assume this is because the amount of Zcash code is going to get bigger soon with orchard and with other changes in `GetAddress`. It makes sense to me to do it this way, but it is unusual to have this sort of arrangement in the present code. Until now the apps in `core/src/apps` did not interact amongst each other, with the exception of `src/apps/common`. However, in this PR the Bitcoin app is calling the Zcash app. @matejcik should approve this arrangement.

File: core/src/apps/zcash/signer.py
Comment (andrewkozlik) [inline core/src/apps/zcash/signer.py:42]: seemless -> seamless

**Comment by andrewkozlik** [comment]
When is the NU5 upgrade going to be activated on Mainnet?

**Comment by jarys** [comment]
> When is the NU5 upgrade going to be activated on Mainnet?

In the end of May 2022 according to the [Zcash timeline](https://z.cash/support/schedule/).

File: core/src/apps/zcash/signer.py
Comment (andrewkozlik) [inline core/src/apps/zcash/signer.py:45]: After discussion with @matejcik, we don't think that changing the version here is a good idea. If the version is not 5, then the signing should fail, because it indicates that the host application is expecting a different transaction version than what Trezor is going to be providing. BTW, note that the warning would only get logged in the debug firmware.

File: core/src/apps/bitcoin/sign_tx/__init__.py
Comment (andrewkozlik) [inline core/src/apps/bitcoin/sign_tx/__init__.py:76]: @matejcik and I are inclined to choosing the signer based on `msg.version` only, i.e. ignore the coin name. Based on past experience we can expect the other Zcashlike coins to adopt NU5 sooner or later.

File: core/src/apps/zcash/sig_hasher.py
Comment (jarys) [reply core/src/apps/zcash/sig_hasher.py:150]: Yay , I didn't know we have a style formater. Ok, I will be using it.

Comment (andrewkozlik) [reply core/src/apps/zcash/sig_hasher.py:150]: More information/guidelines about contributing to Trezor: https://github.com/trezor/trezor-firmware/blob/master/docs/misc/contributing.md.

File: core/src/apps/zcash/sig_hasher.py
Comment (jarys) [reply core/src/apps/zcash/sig_hasher.py:43]: I don't like an idea of passing `tx` every time some digest is needed while using it only once. I will pass it as argument of `create_sig_hasher()`.

File: core/src/apps/zcash/sig_hasher.py
Comment (jarys) [reply core/src/apps/zcash/sig_hasher.py:170]: You are right. They decided to use *scriptPubKey* according to [this comment](https://github.com/zcash/zips/pull/577#discussion_r781645348). 

However notice that calling `write_bip143_script_code_prefixed` has currently only two possible results:
- P2PKH scriptPubKey for multiple possible script types
- multisig scriptPubKey for multisig

The function always return a scriptPubKey so it could be renamed to `write_bip143_script_pubkey_prefixed`. Or even to `write_script_pubkey_prefixed` since it is currently used only in Zcash and it is not further related to the BIP-143.

Or preferably just to `write_script_pubkey` with `prefixed: bool = true` argument for sake of universality.

**Comment by jarys** [comment]
I made a significant change to the code when I was modifying `ZcashSigHasher` to use use script_pubkey instead of script_code.

Note that neither `hash143` nor `hash341` does not receive script_pubkey argument. I considered writing a new function `write_script_pubkey` but then I decided to assimilate the Taproot approach, where script_pubkey is derived by the method `signer.input_derive_script(txi)`.

This led me to unimplementing `ZcashExtendedSigHasher.hash143` and overriding all methods where `hash143` method occured, i.e. `sign_bip143_input` and `get_tx_digest`.

File: core/src/apps/zcash/sig_hasher.py
Comment (jarys) [reply core/src/apps/zcash/sig_hasher.py]: I will keep it for future use cases and for now I will add a comment that this is a dead code.

File: core/src/apps/zcash/sig_hasher.py
Comment (jarys) [reply core/src/apps/zcash/sig_hasher.py]: I

- rename original `ZcashSigHasher` to `Zip243SigHasher`
- rename new `ZcashSigHasher` to `ZcashHasher`
- merge `ZcashExtendedSigHasher` into `ZcashHasher`

in https://github.com/trezor/trezor-firmware/pull/2166/commits/9cec3b7ed08d2c613717db28d6079759ea3c2712

File: core/src/apps/zcash/signer.py
Comment (jarys) [reply core/src/apps/zcash/signer.py:78]: I agree.

File: tests/device_tests/zcash/test_sign_tx.py
Comment (jarys) [reply tests/device_tests/zcash/test_sign_tx.py:104]: I agree. I will submit these transactions as soon as the testnet will be relaunched, which should be in the beginning of April.

File: core/src/apps/zcash/sig_hasher.py
Comment (jarys) [reply core/src/apps/zcash/sig_hasher.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/976d716b73f53089fda19981d27c8be3041c391c and https://github.com/trezor/trezor-firmware/pull/2166/commits/4abffe5215e1db8d972ae1ae29b81d8be4323c5f

File: core/src/apps/zcash/sig_hasher.py
Comment (jarys) [reply core/src/apps/zcash/sig_hasher.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/060aa20c4d7490f209b520230ea2122f93dd4dcb

File: core/src/apps/zcash/sig_hasher.py
Comment (jarys) [reply core/src/apps/zcash/sig_hasher.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/079f0a118983a7fbecc12686c7b78ade78d2f650

File: core/src/apps/zcash/signer.py
Comment (jarys) [reply core/src/apps/zcash/signer.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/4bce23cffea5c3cf4b97e7a0061559cb8ae6e352

File: core/src/apps/zcash/signer.py
Comment (jarys) [reply core/src/apps/zcash/signer.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/4bce23cffea5c3cf4b97e7a0061559cb8ae6e352

File: core/src/apps/bitcoin/sign_tx/__init__.py
Comment (jarys) [reply core/src/apps/bitcoin/sign_tx/__init__.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/71c85f9a8fe9a8a00a3ea779af87b71b1c384ed3

File: core/src/apps/bitcoin/sign_tx/tx_info.py
Comment (andrewkozlik) [inline core/src/apps/bitcoin/sign_tx/tx_info.py:126]: I think there was a good reason for `self.tx = tx` to be assigned here instead of in the parent class. It had to do with type checking. (Note that `make pystyle` also runs a static type-checker.)

File: core/src/apps/bitcoin/sign_tx/__init__.py
Comment (andrewkozlik) [reply core/src/apps/bitcoin/sign_tx/__init__.py]: Also the class name `Zcashlike` -> `ZcashV4`.

Comment (jarys) [reply core/src/apps/bitcoin/sign_tx/__init__.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/7c467f6fd3372d8c184f8b0a3ab18dac1e6e62cc

File: core/src/apps/bitcoin/sign_tx/__init__.py
Comment (jarys) [reply core/src/apps/bitcoin/sign_tx/__init__.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/4b94cad001411de4a84864be9fa678c53f74025d

File: core/src/apps/zcash/signer.py
Comment (jarys) [reply core/src/apps/zcash/signer.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/4b94cad001411de4a84864be9fa678c53f74025d

File: core/src/apps/zcash/signer.py
Comment (jarys) [reply core/src/apps/zcash/signer.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/4b94cad001411de4a84864be9fa678c53f74025d

File: core/src/apps/bitcoin/sign_tx/tx_info.py
Comment (jarys) [reply core/src/apps/bitcoin/sign_tx/tx_info.py:126]: https://github.com/trezor/trezor-firmware/pull/2166/commits/14148f6d577059c8739b6c40e4b746da0b3424d1

File: core/src/apps/bitcoin/sign_tx/tx_info.py
Comment (jarys) [reply core/src/apps/bitcoin/sign_tx/tx_info.py]: https://github.com/trezor/trezor-firmware/pull/2166/commits/a9ef2a93a5e656fd6f5510ce3511502c842b58d3

File: core/src/apps/zcash/signer.py
Comment (andrewkozlik) [inline core/src/apps/zcash/signer.py:45]: Why is this actually needed? Couldn't the sighasher work just as well with `PrevTx`?

**Comment by jarys** [comment]
Nice! All CI checks are passing now.

File: core/src/apps/zcash/signer.py
Comment (jarys) [reply core/src/apps/zcash/signer.py]: Yes, it can. https://github.com/trezor/trezor-firmware/pull/2166/commits/f4fb43e219e29106fd6824772ba243112a76bcc1

I was just lazy to check, because currently we don't need it.

**Review by andrewkozlik — APPROVED** [review]
utACK

TODO:
- @matejcik please see https://github.com/trezor/trezor-firmware/pull/2166#pullrequestreview-920081522, when you have the time.
- When the new testnet activates:
    - on-chain device tests 
    - add multisig device test
- Now:
    - <s>please rebase over master and squash the commits as much as you can (especially let's not clutter the history with "make pystyle", "clean up ...", "fix typo" etc.)</s>
    - <s>push the rebased branch to the `trezor/trezor-firmware` repository to run the ([full](https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines)) CI</s>

**Comment by grdddj** [comment]
Update from creating test vectors:
- we are currently unable to create a `V5` transaction that would be understood by the testnet
  - we have tried https://sochain.com and https://blockbook-dev.corp.sldev.cz:19132 as providers to send the transactions, and in both cases it was rejected
- [grdddj/zcash_v5_testnet_tx_vectors](https://github.com/trezor/trezor-firmware/tree/grdddj/zcash_v5_testnet_tx_vectors) branch contains a `test_spend_real_output` "test", which should generate a valid and announcable transaction, as it spends a valid `UTXO` - https://blockbook-dev.corp.sldev.cz:19132/tx/259f886923e6bf4aa77f5e0c715bb8883be0aa4dd68f482e084af8de1aa20ae7 - owned by `tmQoJ3PTXgQLaRRZZYT6xk8XtjRbr2kCqwu`, which is our address under  `m/44h/1h/0h/0/0`
  - the resulting `tx_hex` is `0500008085202f89b4d0d6c2000000000000000001e70aa21adef84a082e488fd64daae03b88b85b710c5e7fa74abfe62369889f25000000006a4730440220462a4b71f50f473adb0a6c5069db368db3940994cc4020f23764609cd7ef69db0220133e5427b68b95c036235ca01dd8e9f5fd01effa68171320a78ea5845370c6c70121030e669acac1f280d1ddf441cd2ba5e97417bf2689e4bbec86df4f831bf9f7ffd0ffffffff01d0ae3f00000000001976a9141215d421cb8cec1dea62cbd9e4e07c01520d873f88ac000000`
  - putting it into https://blockbook-dev.corp.sldev.cz:19132/sendtx results in `-22: TX decode failed`
  - we have verified that blockbook currently indexes `v5` transactions

**Comment by andrewkozlik** [comment]
@grdddj `version_group_id` needs to be set to `0x26a7270a`, see https://blockbook-dev.corp.sldev.cz:19132/tx/733df241dcfe7f03d4ef0e36450048f09685a42ee734543ffae220c42292eb99

File: core/src/apps/zcash/signer.py
Comment (andrewkozlik) [reply core/src/apps/zcash/signer.py:60]: It works, but I agree, it's not pretty. I was thinking of changing it to `step3_verify_inputs(check_prev_tx: bool)`. In `bitcoin.py` it would be called with `step3_verify_inputs(not self.taproot_only)` and in `Zcash` (v5) we would have:
```py
    async def step3_verify_inputs(check_prev_tx: bool) -> None:
        await super().step3_verify_inputs(False)
```
However, I like @matejcik's proposal just as well, only not sure about the function naming. `verify_inputs_fast() -> verify_inputs_no_prevtx()`?

**Comment by andrewkozlik** [comment]
> @grdddj will create test vectors.

I think we should have the following types of transactions in the device tests:
- v4 UTXO spent in a v5 transaction
- v5 UTXO spent in a v5 transaction
- v5 multisig UTXO spent in a v5 transaction

**Comment by grdddj** [comment]
> `version_group_id` needs to be set to `0x26a7270a`

Nice, thanks for this! Is it the same in every transaction? And what about the `branch_id` value?

**Comment by andrewkozlik** [comment]
> > `version_group_id` needs to be set to `0x26a7270a`
> 
> Nice, thanks for this! Is it the same in every transaction? And what about the `branch_id` value?

I think it should be the same in every transaction. I have no clue where the value comes from. They are not specified in the ZIPs, I just read them from the transactions in the blockchain.

**Comment by grdddj** [comment]
I have force-pushed [grdddj/zcash_v5_testnet_tx_vectors](https://github.com/trezor/trezor-firmware/tree/grdddj/zcash_v5_testnet_tx_vectors) branch, and currently all the test-vector changes are in one commit - https://github.com/trezor/trezor-firmware/commit/1c618d9336f20148358e66ecb627076d281dc8f4

I was mistaken that we already have spend from `v4`, as even that [259f886923e6bf4aa77f5e0c715bb8883be0aa4dd68f482e084af8de1aa20ae7](https://blockbook-dev.corp.sldev.cz:19132/tx/259f886923e6bf4aa77f5e0c715bb8883be0aa4dd68f482e084af8de1aa20ae7) was already `v5`. Luckily, there was one history UTXO still unspent on `tmAgYbANTzZp7YoMkRbbaemQETgV5GkBEjF` ([4b6cecb81c825180786ebe07b65bcc76078afc5be0f1c64e08d764005012380d](https://blockbook-dev.corp.sldev.cz:19132/tx/4b6cecb81c825180786ebe07b65bcc76078afc5be0f1c64e08d764005012380d)), which I could spend.

Connected with spending from a multisig, I also included a test sending funds to multisig, which was a transaction I had to generate anyway.

There are currently two tests marked with `FAKE tx`:
- `test_one_one`, which could be theoretically removed, as we test this in a lot of other places
- `test_external_presigned` - as this is still spending invalid (already spend) inputs. I did not see a big worth recreating it
- the rest (apart from the `pytest.raises(...)`) are valid `Zcash Testnet` transactions

@matejcik also pointed out it would be nice to test spending of `v4` and `v5` in the same transaction, but I could not find any other old `v4` outputs and also failed to generate some myself. Maybe somebody else will be more lucky.

**Comment by andrewkozlik** [comment]
> * `test_one_one`, which could be theoretically removed, as we test this in a lot of other places

I'd remove it, since we already have `test_spend_v4_input` and `test_spend_v5_input`, which do exactly this.

**Comment by grdddj** [comment]
> > * `test_one_one`, which could be theoretically removed, as we test this in a lot of other places
> 
> I'd remove it, since we already have `test_spend_v4_input` and `test_spend_v5_input`, which do exactly this.

Removed in 1a97288 - a fixup commit to the above one

**Comment by grdddj** [comment]
One more fixup - bf64793991e9c0f989d6ab2ad3c36b342ef67d53 - to fix the `fixtures.json` expected hashes. I made the mistake of generating the hashes with enabled animations. CI was correctly showing `UI` mismatch

**Comment by matejcik** [comment]
will be merged in #2245

**Pull request closed without merging**
